### ISSUE TITLE
Stack overflow with rayon::scope and spawn

### ISSUE THREAD

**Issue by mdesmedt** [rayon-rs/rayon#854]
Here is a simple piece of code which causes a stack overflow with rayon 1.5.0:
```
use rayon::prelude::*;

fn do_work(index: u32) {
    if index == 123015453 {
        println!("Hello world");
    }
}

fn main() {
    rayon::scope(|s| {
        let num_blocks: u32 = 10000;
        for block in 0..num_blocks {
            s.spawn(move |_| {
                let num_pixels: u32 = block;
                (0..num_pixels).into_par_iter().for_each(|index| {
                    do_work(index);
                }); // par_iter over pixels
            }); // spawn
        } // loop blocks
    }); // scope
}
```

The callstack repeats this pattern:
```
...
static void rayon_core::registry::WorkerThread::execute(struct rayon_core::job::JobRef)
void rayon_core::registry::WorkerThread::wait_until_cold(struct rayon_core::latch::CoreLatch *)
void rayon_core::registry::WorkerThread::wait_until<rayon_core::latch::SpinLatch>(struct rayon_core::latch::SpinLatch *)
void rayon_core::join::join_context::{{closure}}<closure-0,closure-1,tuple<>,tuple<>>(struct rayon_core::join::join_context::closure-0, struct rayon_core::registry::WorkerThread *, bool)
void rayon_core::registry::in_worker<closure-0,tuple<tuple<>, tuple<>>>(struct rayon_core::join::join_context::closure-0)
void rayon_core::join::join_context<closure-0,closure-1,tuple<>,tuple<>>(struct rayon::iter::plumbing::bridge_producer_consumer::helper::closure-0, struct rayon::iter::plumbing::bridge_producer_consumer::helper::closure-1)
static void rayon::iter::plumbing::bridge_producer_consumer::helper<rayon::range::IterProducer<u32>,rayon::iter::for_each::ForEachConsumer<closure-0>>(unsigned __int64, bool, struct rayon::iter::plumbing::LengthSplitter, struct rayon::range::IterProducer<u32>, struct rayon::iter::for_each::ForEachConsumer<closure-0>)
void rayon::iter::plumbing::bridge_producer_consumer::helper::{{closure}}<rayon::range::IterProducer<u32>,rayon::iter::for_each::ForEachConsumer<closure-0>>(struct rayon::iter::plumbing::bridge_producer_consumer::helper::closure-1, struct rayon_core::FnContext)
void rayon_core::join::join_context::call_b::{{closure}}<tuple<>,closure-1>(struct rayon_core::join::join_context::call_b::closure-0, bool)
void rayon_core::job::StackJob<rayon_core::latch::SpinLatch, closure-0, tuple<>>::run_inline<rayon_core::latch::SpinLatch,closure-0,tuple<>>(struct rayon_core::job::StackJob<rayon_core::latch::SpinLatch, closure-0, tuple<>>, bool)
void rayon_core::join::join_context::{{closure}}<closure-0,closure-1,tuple<>,tuple<>>(struct rayon_core::join::join_context::closure-0, struct rayon_core::registry::WorkerThread *, bool)
void rayon_core::registry::in_worker<closure-0,tuple<tuple<>, tuple<>>>(struct rayon_core::join::join_context::closure-0)
void rayon_core::join::join_context<closure-0,closure-1,tuple<>,tuple<>>(struct rayon::iter::plumbing::bridge_producer_consumer::helper::closure-0, struct rayon::iter::plumbing::bridge_producer_consumer::helper::closure-1)
static void rayon::iter::plumbing::bridge_producer_consumer::helper<rayon::range::IterProducer<u32>,rayon::iter::for_each::ForEachConsumer<closure-0>>(unsigned __int64, bool, struct rayon::iter::plumbing::LengthSplitter, struct rayon::range::IterProducer<u32>, struct rayon::iter::for_each::ForEachConsumer<closure-0>)
static void rayon::iter::plumbing::bridge_producer_consumer<rayon::range::IterProducer<u32>,rayon::iter::for_each::ForEachConsumer<closure-0>>(unsigned __int64, struct rayon::range::IterProducer<u32>, struct rayon::iter::for_each::ForEachConsumer<closure-0>)
void rayon::iter::plumbing::bridge::{{impl}}::callback<rayon::iter::for_each::ForEachConsumer<closure-0>,u32,rayon::range::IterProducer<u32>>(struct rayon::iter::plumbing::bridge::Callback<rayon::iter::for_each::ForEachConsumer<closure-0>>, struct rayon::range::IterProducer<u32>)
void rayon::range::{{impl}}::with_producer<rayon::iter::plumbing::bridge::Callback<rayon::iter::for_each::ForEachConsumer<closure-0>>>(struct rayon::range::Iter<u32>, struct rayon::iter::plumbing::bridge::Callback<rayon::iter::for_each::ForEachConsumer<closure-0>>)
void rayon::iter::plumbing::bridge<rayon::range::Iter<u32>,rayon::iter::for_each::ForEachConsumer<closure-0>>(struct rayon::range::Iter<u32>, struct rayon::iter::for_each::ForEachConsumer<closure-0>)
void rayon::range::{{impl}}::drive_unindexed<rayon::iter::for_each::ForEachConsumer<closure-0>>(struct rayon::range::Iter<u32>, struct rayon::iter::for_each::ForEachConsumer<closure-0>)
rayon::iter::for_each::for_each::h15817c5d0f369c94
rayon::iter::ParallelIterator::for_each::h26990b208229649d
void rayon_stack_overflow::main::{{closure}}::{{closure}}(struct rayon_stack_overflow::main::{{closure}}::closure-0, struct rayon_core::scope::Scope *)
void rayon_core::scope::{{impl}}::spawn::{{closure}}::{{closure}}<closure-0>(struct rayon_core::scope::{{impl}}::spawn::{{closure}}::closure-0)
void std::panic::{{impl}}::call_once<tuple<>,closure-0>(struct std::panic::AssertUnwindSafe<closure-0>)
static void std::panicking::try::do_call<std::panic::AssertUnwindSafe<closure-0>,tuple<>>(unsigned char *)
1400047C7 (@7ff7fab547c7..7ff7fab5482d:3)
union core::result::Result<tuple<>, alloc::boxed::Box<Any, alloc::alloc::Global>> std::panicking::try<tuple<>,std::panic::AssertUnwindSafe<closure-0>>(struct std::panic::AssertUnwindSafe<closure-0>)
union core::result::Result<tuple<>, alloc::boxed::Box<Any, alloc::alloc::Global>> std::panic::catch_unwind<std::panic::AssertUnwindSafe<closure-0>,tuple<>>(struct std::panic::AssertUnwindSafe<closure-0>)
union core::result::Result<tuple<>, alloc::boxed::Box<Any, alloc::alloc::Global>> rayon_core::unwind::halt_unwinding<closure-0,tuple<>>(struct rayon_core::scope::{{impl}}::spawn::{{closure}}::closure-0)
static core::option::Option rayon_core::scope::ScopeBase::execute_job_closure<closure-0,tuple<>>(struct rayon_core::scope::{{impl}}::spawn::{{closure}}::closure-0)
static void rayon_core::scope::ScopeBase::execute_job<closure-0>(struct rayon_core::scope::{{impl}}::spawn::{{closure}}::closure-0)
void rayon_core::scope::{{impl}}::spawn::{{closure}}<closure-0>(struct rayon_core::scope::{{impl}}::spawn::closure-0)
static void rayon_core::job::{{impl}}::execute<closure-0>(struct rayon_core::job::HeapJob<closure-0> *)
void rayon_core::job::JobRef::execute()
static void rayon_core::registry::WorkerThread::execute(struct rayon_core::job::JobRef)
...
```

I'm fairly new to Rust and rayon so it might be what I'm doing is wrong, but as my code isn't obviously recursing I wouldn't expect rayon internally to recurse either. This behavior poses a danger to users of the lib as it might work in trivial cases but panic in others.

**Comment by cuviper** [comment]
Rayon has implicit "recursion" due to work stealing. That is, whenever a rayon thread is blocked on the result from another rayon thread, it will look for other pending work to do in the meantime. That stolen work is executed directly from the same stack where it was blocked.

Your `par_iter().for_each()` becomes a bunch of nested `join`s, and each one of those may block if one half gets stolen to a new thread. Since stealing is somewhat random, the pool will have a mix of stolen `join`s and new `spawn`s creating even more `join`s, and I can definitely see how that might get out of control. You're not doing anything *wrong*, but I'm not sure how to tame that.

The new `in_place_scope` in #844 would probably help, assuming you start from outside the thread pool. Then each `spawn` will go into the pool's external queue, whereas `scope` runs in the pool and pushes `spawn`s into a thread's local queue. The local thread queues have priority over the external queue when it comes to work stealing, so that would let you prioritize finishing current `par_iter`s before starting new `spawn`s, for a bit less of a task explosion.

**Comment by cuviper** [comment]
Maybe we could also have some sort of `spawn_external` as a way of indicating lower priority, where we push to the external queue even if you call this from within the thread pool.

**Comment by cuviper** [comment]
>         for block in 0..num_blocks {
>             s.spawn(move |_| {
>                 let num_pixels: u32 = block;
>                 (0..num_pixels).into_par_iter().for_each(|index| {
>                     do_work(index);
>                 }); // par_iter over pixels
>             }); // spawn
>         } // loop blocks

BTW, what do you mean by this "loop blocks" comment? The `spawn`s do not wait for completion here, so the loop will queue up all `num_blocks` iterations in rapid succession. Only the end of the `scope` will block for all of the `spawn`s.

**Comment by mdesmedt** [comment]
> > ```
> >     for block in 0..num_blocks {
> >         s.spawn(move |_| {
> >             let num_pixels: u32 = block;
> >             (0..num_pixels).into_par_iter().for_each(|index| {
> >                 do_work(index);
> >             }); // par_iter over pixels
> >         }); // spawn
> >     } // loop blocks
> > ```
> 
> BTW, what do you mean by this "loop blocks" comment? The `spawn`s do not wait for completion here, so the loop will queue up all `num_blocks` iterations in rapid succession. Only the end of the `scope` will block for all of the `spawn`s.

Hah. It's not a "block" in terms of concurrency :). I derived this snippet from my little "[Raytracing in one weekend](https://github.com/mdesmedt/rust_one_weekend/blob/20e563772acff1f6efbc1d5b57e532a295262b84/src/render.rs#L183)" implementation. Where I generate a list of "renderblocks" to render in FIFO order. Complete overkill really for a bunch of spheres, but it looks pretty.

Thanks for your explanation. It seems like this potential for stack explosions is inherent to the current work stealing architecture of rayon. As I hit this in my code "in production" so to speak a mitigation or alternative way of expressing the work so hitting this won't happen would be great.

**Comment by cuviper** [comment]
Oh of course, you mean each "block" in `0..num_blocks`. :sweat_smile:

**Comment by konstin** [comment]
We're hitting this problem in uv: https://github.com/astral-sh/uv/issues/12769, stack trace: https://gist.github.com/konstin/ff30565563675ca8a4b55e4258a7b511. In another case the stack was so large that windbg couldn't show it with `k`.

The code that triggers the stack overflow is this: https://github.com/astral-sh/uv/blob/c94777fc544ef53688cc20ffeaf16c45ef6290c7/crates/uv-extract/src/sync.rs#L12-L89, which we run in parallel, in the example case for ~400 unzips.

Is the workaround with `in_place_scope`/`spawn` over `into_par_iter` still applicable?

**Comment by cuviper** [comment]
> Is the workaround with `in_place_scope`/`spawn` over `into_par_iter` still applicable?

Yes, although that's not going to be perfect either, just a crude form of prioritization. Maybe it's good enough though.

With ~400 unzips in play, I wonder whether that inner `into_par_iter` is really helping much in the first place? You could just work with coarse parallelism at the outer level, and you'd only lose some potential at the tail end when there are only a few left.

**Comment by teor2345** [comment]
> Your `par_iter().for_each()` becomes a bunch of nested `join`s, and each one of those may block if one half gets stolen to a new thread. Since stealing is somewhat random, the pool will have a mix of stolen `join`s and new `spawn`s creating even more `join`s, and I can definitely see how that might get out of control. You're not doing anything _wrong_, but I'm not sure how to tame that.
> 
> The new `in_place_scope` in [#844](https://github.com/rayon-rs/rayon/pull/844) would probably help, assuming you start from outside the thread pool. Then each `spawn` will go into the pool's external queue, whereas `scope` runs in the pool and pushes `spawn`s into a thread's local queue. The local thread queues have priority over the external queue when it comes to work stealing, so that would let you prioritize finishing current `par_iter`s before starting new `spawn`s, for a bit less of a task explosion.

We have a similar issue with explicit `rayon` recursion using `join` in this `rust-kzg` code:
https://github.com/grandinetech/rust-kzg/blob/20ab11604e7c6747f0bc8ab404d324c35ef74fed/blst/src/fft_fr.rs#L40-L46

More details here:
https://github.com/autonomys/subspace/issues/3698#issuecomment-3328866026

We end up with a stack overflow due to 600+ recursions on the same stack of:  
```
#1880 0x000055cce0a425d4 in rust_kzg_blst::fft_fr::fft_fr_fast ()                                                                              
#1881 0x000055cce0a2c39c in rayon_core::job::StackJob<L,F,R>::run_inline ()           
#1882 0x000055cce0a30f5b in rayon_core::join::join_context::{{closure}} () 
#1883 0x000055cce0a425d4 in rust_kzg_blst::fft_fr::fft_fr_fast ()         
```
                                      
There appears to be work stealing going on, because at the top of the stack we have:                
```
#5  0x000055cce0b32ea2 in std::sys::pal::unix::stack_overflow::imp::signal_handler ()                                                          
#6  <signal handler called>                                                                                                                    
#7  0x000055cce06dc9d1 in crossbeam_epoch::internal::Global::collect ()                                                                        
#8  0x000055cce093ec50 in crossbeam_epoch::default::with_handle ()                                                                             
#9  0x000055cce093ee7d in crossbeam_deque::deque::Stealer<T>::steal ()                                                                         
#10 0x000055cce093a825 in rayon_core::registry::WorkerThread::wait_until_cold ()                                                               
#11 0x000055cce0a30fb8 in rayon_core::join::join_context::{{closure}} ()                                                                       
#12 0x000055cce0a425d4 in rust_kzg_blst::fft_fr::fft_fr_fast ()                                                                                
```

Unfortunately this only happens occasionally, and it's in code we don't control (both `rayon` and `rust-kzg` are external dependencies).

I'm not sure if the `in_place_scope` solution would work for us, because this is happening on a rayon thread:
```
#1924 0x000055cce110ac9c in rayon_core::join::join_context::{{closure}} ()
#1925 0x000055cce0f20727 in rayon::iter::plumbing::bridge_producer_consumer::helper ()
#1926 0x000055cce0f2dc08 in <rayon_core::job::StackJob<L,F,R> as rayon_core::job::Job>::execute ()
#1927 0x000055cce093b7dd in rayon_core::registry::WorkerThread::wait_until_cold ()
#1928 0x000055cce0939bb5 in rayon_core::registry::ThreadBuilder::run ()
#1929 0x000055cce09411d6 in std::sys::backtrace::__rust_begin_short_backtrace ()
#1930 0x000055cce0940594 in core::ops::function::FnOnce::call_once{{vtable.shim}} ()
#1931 0x000055cce0b495a7 in std::sys::pal::unix::thread::Thread::new::thread_start ()
```

Ideally, we'd like this to be fixed in our dependencies, possibly by (in rough order of preference):
- `rayon` implementing work stealing using iteration rather than recursion
- once a `rayon` thread stack gets too deep, it doesn't steal any more work
- limiting work stealing to 500 jobs total in the entire process
- limiting the number of `rayon` jobs to 500 in the entire process (this would be fine for us, because almost all machines we run on have ~64 logical cores or less)
-

**Comment by cuviper** [comment]
> One thing you could do is reduce the amount of recursion inside rayon. So, for example, in:
> 
> ```
> rayon_core::job::StackJob<L,F,R>::run_inline ()           
> rayon_core::join::join_context::{{closure}} () 
> ```
> 
> Maybe `run_inline` could return the user function to be called, and the `join_context::{{closure}}` could call that function? Then there would be 1 rayon stack frame added every time work was stolen, rather than 2.

If optimization is enabled, this kind of thing should be inlined already. Or are you having trouble with debug builds?

> The stack overflow occurs when threads repeatedly steal work from other threads, and repeatedly add 12 jobs from each different task onto their stack. (Or in the worst case scenario, 11 frames from one branch of a task, then 10 from the next branch of the same task, then 9 from the next branch, and so on.)
> 
> If we stopped work stealing, this explosion of stack frames wouldn't happen.

It at least still needs to "steal" from its local queue, as `join` does with `take_local_job`, but it couldn't steal from higher up its own stack or else you have the same explosion from expanding a different branch of the job tree.

This also gets messy if there are any other tasks mixed in the queue, like a `spawn` from some outer `scope` -- which doesn't *need* to block the current `join`, but we don't have a way to de-queue the `job_b` otherwise. Essentially, this is a composition problem, where we don't have enough separation of concerns -- and *generally* this keeps things simple!

It would be nice to solve that, because there are other areas where "stop work stealing" could come in handy, like enabling a "critical section" when you're holding a `Mutex` guard. I don't have confidence that it can be done safely/reliably though.

**Comment by teor2345** [comment]
> > One thing you could do is reduce the amount of recursion inside rayon. So, for example, in:
> > ```
> > rayon_core::job::StackJob<L,F,R>::run_inline ()           
> > rayon_core::join::join_context::{{closure}} () 
> > ```
> > 
> > Maybe `run_inline` could return the user function to be called, and the `join_context::{{closure}}` could call that function? Then there would be 1 rayon stack frame added every time work was stolen, rather than 2.
> 
> If optimization is enabled, this kind of thing should be inlined already. Or are you having trouble with debug builds?

Optimisation is enabled, including LTO. Here are our exact optimisation settings:
https://github.com/autonomys/subspace/blob/b59b800bcb82917c40ff293ffb2d9b15efd47796/Cargo.toml#L362-L364

Unfortunately these functions/closures are not being inlined on x86_64.

I added line tables to diagnose this issue - do you think that's stopping the inlining?
Here's the exact command I used:
`RUSTFLAGS="-C debuginfo=line-tables-only -C split-debuginfo=packed" cargo build --locked --profile production --bin subspace-node --bin subspace-farmer`

> > The stack overflow occurs when threads repeatedly steal work from other threads, and repeatedly add 12 jobs from each different task onto their stack. (Or in the worst case scenario, 11 frames from one branch of a task, then 10 from the next branch of the same task, then 9 from the next branch, and so on.)
> > If we stopped work stealing, this explosion of stack frames wouldn't happen.
> 
> It at least still needs to "steal" from its local queue, as `join` does with `take_local_job`, but it couldn't steal from higher up its own stack or else you have the same explosion from expanding a different branch of the job tree.
> 
> This also gets messy if there are any other tasks mixed in the queue, like a `spawn` from some outer `scope` -- which doesn't _need_ to block the current `join`, but we don't have a way to de-queue the `job_b` otherwise. Essentially, this is a composition problem, where we don't have enough separation of concerns -- and _generally_ this keeps things simple!

We shouldn't have any other tasks mixed in, but I'll check that we're creating a separate rayon thread pool for this type of work.

One option would be splitting that thread pool into multiple thread pools, and limiting how many top-level jobs get scheduled on each. 

> It would be nice to solve that, because there are other areas where "stop work stealing" could come in handy, like enabling a "critical section" when you're holding a `Mutex` guard. I don't have confidence that it can be done safely/reliably though.

Ah, that's unfortunate.

**Comment by teor2345** [comment]
Just a note that rewriting the recursive code we're depending on gets us about 5-10% more stack frames before overflow:
https://github.com/grandinetech/rust-kzg/pull/310

But we needed to also optimise for code size to significantly reduce or eliminate the stack overflows:
https://github.com/autonomys/subspace/pull/3716/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R362
